test/gtest/common/zip.cpp

```diff
@@ -4,176 +4,206 @@
 
 #include "xhp-tests.hpp"
 
+template <typename... Rs> auto test_zip(Rs &&...rs) {
+  return xhp::views::zip(std::forward<Rs>(rs)...);
+}
+
 // Fixture
 template <typename T> class Zip : public testing::Test {
 public:
 };
 
 TYPED_TEST_SUITE(Zip, AllTypes);
 
-// Try 1, 2, 3 to check pair/tuple issues
-TYPED_TEST(Zip, Local1) {
+TYPED_TEST(Zip, Dist1) {
   Ops1<TypeParam> ops(10);
 
-  EXPECT_EQ(rng::views::zip(ops.vec), xhp::views::zip(ops.vec));
+  auto local = rng::views::zip(ops.vec);
+  auto dist = test_zip(ops.dist_vec);
+  static_assert(compliant_view<decltype(dist)>);
+  EXPECT_TRUE(check_view(local, dist));
 }
 
-TYPED_TEST(Zip, Local2) {
+TYPED_TEST(Zip, Dist2) {
   Ops2<TypeParam> ops(10);
 
-  EXPECT_EQ(rng::views::zip(ops.vec0, ops.vec1),
-            xhp::views::zip(ops.vec0, ops.vec1));
+  auto local = rng::views::zip(ops.vec0, ops.vec1);
+  auto dist = test_zip(ops.dist_vec0, ops.dist_vec1);
+  static_assert(compliant_view<decltype(dist)>);
+  EXPECT_TRUE(check_view(local, dist));
 }
 
-TYPED_TEST(Zip, Local3) {
+TYPED_TEST(Zip, Dist3) {
   Ops3<TypeParam> ops(10);
 
-  EXPECT_EQ(rng::views::zip(ops.vec0, ops.vec1, ops.vec2),
-            xhp::views::zip(ops.vec0, ops.vec1, ops.vec2));
+  auto local = rng::views::zip(ops.vec0, ops.vec1, ops.vec2);
+  auto dist = test_zip(ops.dist_vec0, ops.dist_vec1, ops.dist_vec2);
+  static_assert(compliant_view<decltype(dist)>);
+  EXPECT_TRUE(check_view(local, dist));
 }
 
-TYPED_TEST(Zip, Local3Distance) {
+TYPED_TEST(Zip, Dist3Distance) {
   Ops3<TypeParam> ops(10);
 
-  EXPECT_EQ(rng::distance(rng::views::zip(ops.vec0, ops.vec1, ops.vec2)),
-            rng::distance(xhp::views::zip(ops.vec0, ops.vec1, ops.vec2)));
+  EXPECT_EQ(
+      rng::distance(rng::views::zip(ops.vec0, ops.vec1, ops.vec2)),
+      rng::distance(test_zip(ops.dist_vec0, ops.dist_vec1, ops.dist_vec2)));
 }
 
-TYPED_TEST(Zip, Local2Mutate) {
-  Ops2<TypeParam> rops(10);
-  Ops2<TypeParam> xops(10);
+TYPED_TEST(Zip, RangeSegments) {
+  Ops1<TypeParam> ops(10);
 
-  auto r = rng::views::zip(rops.vec0, rops.vec1);
-  auto x = rng::views::zip(xops.vec0, xops.vec1);
-  auto n2 = [](auto v) {
-    std::get<0>(v) += 1;
-    std::get<1>(v) = -std::get<1>(v);
-  };
-  rng::for_each(r, n2);
-  rng::for_each(x, n2);
-  EXPECT_EQ(rops.vec0, xops.vec0);
-  EXPECT_EQ(rops.vec1, xops.vec1);
+  auto local = rng::views::zip(ops.vec);
+  auto dist = test_zip(ops.dist_vec);
+  auto flat = rng::views::join(dr::ranges::segments(dist));
+  EXPECT_TRUE(is_equal(local, flat));
 }
 
-TYPED_TEST(Zip, Dist1) {
+#if 0
+// Will not compile with SHP
```

Comment:
@BenBrock. This is a new test that I am disabling because it will not compile with SHP. It looks like SHP zip_view does provide segments from an iterator. It is not urgent, but would be good to fix.